qiskit/algorithms/optimizers/optimizer.py

```diff
@@ -12,15 +12,93 @@
 
 """Optimizer interface"""
 
-from typing import Dict, Any
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List
 
+import warnings
 from enum import IntEnum
 import logging
 from abc import ABC, abstractmethod
 import numpy as np
 
+from qiskit.algorithms.algorithm_result import AlgorithmResult
+
 logger = logging.getLogger(__name__)
 
+POINT = Union[float, np.ndarray]
+
+
+class OptimizerResult(AlgorithmResult):
+    """The result of an optimization routine."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._x = None  # pylint: disable=invalid-name
+        self._fun = None
+        self._jac = None
+        self._nfev = None
+        self._njev = None
+        self._nit = None
+
+    @property
+    def x(self) -> Optional[POINT]:
+        """The final point of the minimization."""
+        return self._x
+
+    @x.setter
+    def x(self, x: Optional[POINT]) -> None:
+        """Set the final point of the minimization."""
+        self._x = x
+
+    @property
+    def fun(self) -> Optional[float]:
+        """The final value of the minimization."""
+        return self._fun
+
+    @fun.setter
+    def fun(self, fun: Optional[float]) -> None:
+        """Set the final value of the minimization."""
+        self._fun = fun
+
+    @property
+    def jac(self) -> Optional[POINT]:
+        """The final gradient of the minimization."""
+        return self._jac
+
+    @jac.setter
+    def jac(self, jac: Optional[POINT]) -> None:
+        """Set the final gradient of the minimization."""
+        self._jac = jac
+
+    @property
+    def nfev(self) -> Optional[int]:
+        """The total number of function evaluations."""
+        return self._nfev
+
+    @nfev.setter
+    def nfev(self, nfev: Optional[int]) -> None:
+        """Set the total number of function evaluations."""
+        self._nfev = nfev
+
+    @property
+    def njev(self) -> Optional[int]:
+        """The total number of gradient evaluations."""
+        return self._njev
+
+    @njev.setter
+    def njev(self, njev: Optional[int]) -> None:
+        """Set the total number of gradient evaluations."""
+        self._njev = njev
+
+    @property
+    def nit(self) -> Optional[int]:
```

Comment:
I don't think `nit` is intuitive. Why not sort of `num_iterations`?

Reply:
Because that's what scipy uses in their result object: https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html#scipy.optimize.OptimizeResult

Reply:
Since we're modeling the optimizers close to SciPy having the same attribute names is probably useful, since the algorithms can then process both result objects 🙂 (but I agree it's not a very good name)

Reply:
I see. It makes sense to use the same name as scipy. I didn't notice scipy naming.